src/main/java/net/wasdev/wlp/common/plugins/util/DevUtil.java

```diff
@@ -0,0 +1,974 @@
+            serverEnvFile = new File(serverDirectory.getCanonicalPath() + "/server.env");
+            if (serverEnvBackup.exists()) {
+                // Restore original server.env file
+                try {
+                    Files.copy(serverEnvBackup.toPath(), serverEnvFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                } catch (IOException e) {
+                    error("Could not restore server.env: " + e.getMessage());
+                }
+                serverEnvBackup.delete();
+            } else {
+                // Delete server.env file
+                serverEnvFile.delete();
+            }
+        } catch (IOException e) {
+            error("Could not retrieve server.env: " + e.getMessage());
+        }
+    }
+    
+    public void cleanUpTempConfig() {
+        if (this.tempConfigPath != null){
+            File tempConfig = this.tempConfigPath.toFile();
+            if (tempConfig.exists()){
+                try {
+                    FileUtils.deleteDirectory(tempConfig);
+                    debug("Sucessfully deleted liberty:dev temporary configuration folder");
+                } catch (IOException e) {
+                    error("Could not delete liberty:dev temporary configuration folder");
+                }
+            }
+        }
+    }
+
+    public void addShutdownHook(final ThreadPoolExecutor executor) {
+        // shutdown hook to stop server when x mode is terminated
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                debug("Inside Shutdown Hook, shutting down server");
+                
+                cleanUpTempConfig();
+                cleanUpServerEnv();
+
+                if (hotkeyReader != null) {
+                    hotkeyReader.shutdown();
+                }
+
+                // shutdown tests
+                executor.shutdown();
+
+                // stopping server
+                stopServer();
+            }
+        });
+    }
+
+    public void enableServerDebug(int libertyDebugPort) throws IOException {
+        String serverEnvPath = serverDirectory.getCanonicalPath() + "/server.env";
+        File serverEnvFile = new File(serverEnvPath);
+        StringBuilder sb = new StringBuilder();
+        if (serverEnvFile.exists()) {
+            debug("server.env already exists");
+            File serverEnvBackup = new File(serverEnvPath + ".bak");
+            Files.copy(serverEnvFile.toPath(), serverEnvBackup.toPath());
+            boolean deleted = serverEnvFile.delete();
+
+            if (!deleted) {
+                error("Could not move existing liberty:dev server.env file");
+            }
+
+            BufferedReader reader = new BufferedReader(new FileReader(serverEnvBackup));
+            String line;
+            while ((line = reader.readLine()) != null) {
+                sb.append(line);
+                sb.append("\n");
+            }
+            reader.close();
```

Comment:
Anywhere you are using BufferedReader/BufferedWriter or any input stream or output stream, you should wrap the use of it in a try/finally and call the close() method in the finally. This will ensure they do not get left open.

Reply:
fixed in https://github.com/WASdev/ci.common/pull/52